File: ruby/translator.rb
# Braille mappings for English alphabet and digits
BRAILLE_TO_ENGLISH = {
  'O.....' => 'a', 'O.O...' => 'b', 'OO....' => 'c', 'OO.O..' => 'd', 'O..O..' => 'e',
  'OOO...' => 'f', 'OOOO..' => 'g', 'O.OO..' => 'h', '.OO...' => 'i', '.OOO..' => 'j',
  'O...O.' => 'k', 'O.O.O.' => 'l', 'OO..O.' => 'm', 'OO.OO.' => 'n', 'O..OO.' => 'o',
  'OOO.O.' => 'p', 'OOOOO.' => 'q', 'O.OOO.' => 'r', '.OO.O.' => 's', '.OOOO.' => 't',
  'O...OO' => 'u', 'O.O.OO' => 'v', '.OOO.O' => 'w', 'OO..OO' => 'x', 'OO.OOO' => 'y',
  'O..OOO' => 'z', '......' => ' ' # only map letters and space here
}

# Numbers map using the same symbols as 'a' to 'j' but require NUMBER_SYMBOL context
BRAILLE_NUMBERS = {
  'O.....' => '1', 'O.O...' => '2', 'OO....' => '3', 'OO.O..' => '4', 'O..O..' => '5',
  'OOO...' => '6', 'OOOO..' => '7', 'O.OO..' => '8', '.OO...' => '9', '.OOO..' => '0'
}

ENGLISH_TO_BRAILLE = BRAILLE_TO_ENGLISH.invert

# Special symbols
CAPITAL_SYMBOL = '.....O'
NUMBER_SYMBOL = '.O.OOO'

def is_braille?(input)
  input.match?(/\A[O.]+\z/)
end

def translate_to_braille(text)
  result = []
  text.each_char do |char|
    if char =~ /[A-Z]/
      result << CAPITAL_SYMBOL
      result << ENGLISH_TO_BRAILLE[char.downcase]
    elsif char =~ /\d/
      result << NUMBER_SYMBOL
      result << ENGLISH_TO_BRAILLE[(char.to_i + 96).chr] # Convert number to corresponding letter a-j for Braille
    else
      result << ENGLISH_TO_BRAILLE[char]
    end
  end
  result.join  # Join without spaces to match the format
end

def translate_to_english(braille)
  result = []
  is_capital = false
  is_number = false

  braille.scan(/.{6}/).each do |symbol|
    if symbol == CAPITAL_SYMBOL
      is_capital = true
    elsif symbol == NUMBER_SYMBOL
      is_number = true
    elsif is_number
      result << BRAILLE_NUMBERS[symbol]
    elsif is_capital
      result << BRAILLE_TO_ENGLISH[symbol].upcase
      is_capital = false
    else
      result << BRAILLE_TO_ENGLISH[symbol]
    end
  end
  result.join
end

if __FILE__ == $0
  
  input_str = ARGV.join(' ')

  if is_braille?(input_str)
    
    puts translate_to_english(input_str)
  else
   
    puts translate_to_braille(input_str)
  end
end

File: javascript/translator.js
// mappings
const brailleAlphabet = {
    'a': 'O.....', 'b': 'O.O...', 'c': 'OO....', 'd': 'OO.O..', 'e': 'O..O..',
    'f': 'OOO...', 'g': 'OOOO..', 'h': 'O.OO..', 'i': '.OO...', 'j': '.OOO..',
    'k': 'O...O.', 'l': 'O.O.O.', 'm': 'OO..O.', 'n': 'OO.OO.', 'o': 'O..OO.',
    'p': 'OOO.O.', 'q': 'OOOOO.', 'r': 'O.OOO.', 's': '.OO.O.', 't': '.OOOO.',
    'u': 'O...OO', 'v': 'O.O.OO', 'w': '.OOO.O', 'x': 'OO..OO', 'y': 'OO.OOO',
    'z': 'O..OOO', ' ': '......', 
    '1': 'O.....', '2': 'O.O...', '3': 'OO....', '4': 'OO.O..', '5': 'O..O..',
    '6': 'OOO...', '7': 'OOOO..', '8': 'O.OO..', '9': '.OO...', '0': '.OOO..'
  };
  
  const brailleSpecial = {
    capital: '.....O',
    number: '.O.OOO'
  };
  

  const englishFromBraille = Object.fromEntries(
    Object.entries(brailleAlphabet).map(([key, value]) => [value, key])
  );
  
  function isBraille(input) {
    return /^[O.]+$/.test(input);
  }
  
  //  English to Braille
  function englishToBraille(input) {
    let result = '';
    let isNumber = false;
  
    for (let char of input) {
      if (char >= 'A' && char <= 'Z') {
        result += brailleSpecial.capital + brailleAlphabet[char.toLowerCase()];
      } else if (char >= '0' && char <= '9') {
        if (!isNumber) {
          result += brailleSpecial.number;
          isNumber = true;
        }
        result += brailleAlphabet[char];
      } else {
        result += brailleAlphabet[char];
        if (char === ' ') {
          isNumber = false;
        }
      }
    }
  
    return result;
  }
  
  // Braille to English
  function brailleToEnglish(input) {
    let result = '';
    let i = 0;
    let isCapital = false;
    let isNumber = false;
  
    while (i < input.length) {
      let brailleChar = input.slice(i, i + 6);
  
      if (brailleChar === brailleSpecial.capital) {
        isCapital = true;
        i += 6;
        continue;
      } else if (brailleChar === brailleSpecial.number) {
        isNumber = true;
        i += 6;
        continue;
      }
  
      let char = englishFromBraille[brailleChar];
  
      if (isCapital) {
        char = char.toUpperCase();
        isCapital = false;
      }
  
      if (isNumber) {
        isNumber = false;
      }
  
      result += char;
      i += 6;
    }
  
    return result;
  }
  

  function translate(input) {
    if (isBraille(input)) {
      return brailleToEnglish(input);
    } else {
      return englishToBraille(input);
    }
  }
  

  const input = process.argv.slice(2).join(' ');
  console.log(translate(input));
  